arcconfig/arcconfig.go

```diff
@@ -0,0 +1,150 @@
+type Operation struct {
+	// Name of the operation. Unique within a APIRuntimeControlConfig.
+	Name string `yaml:"name" json:"name"`
+
+	// Authentication defines the AuthenticationRequirement
+	Authentication AuthenticationRequirement `yaml:"authentication,omitempty" json:"authentication,omitempty"`
+
+	// ConsumerAuthorization defines the ConsumerAuthorization
+	ConsumerAuthorization ConsumerAuthorization `yaml:"consumer_authorization,omitempty" json:"consumer_authorization,omitempty"`
+
+	// HTTP matching rules for this operation.
+	HTTPMatches []HTTPMatch `yaml:"http_match,omitempty" json:"http_match,omitempty"`
+
+	// Name of the target server for this operation. This will be sent to Envoy
+	// for routing to the corresponding upstream cluster upon a successful
+	// authorization of the operation.
+	Target string `yaml:"target" json:"target"`
+}
+
+// AuthenticationRequirement defines the authentication requirement.
+// Precisely one of JWT, Any and All should be set.
+type AuthenticationRequirement struct {
+	// JWT defines the JWTAuthentication
+	JWT JWTAuthentication `yaml:"jwt,omitempty" json:"jwt,omitempty"`
+
+	// Any contains a list of AuthenticationRequirements.
+	// A successful authentication requires one of them being satisfied.
+	Any []AuthenticationRequirement `yaml:"any,omitempty" json:"any,omitempty"`
+
+	// All contains a list of AuthenticationRequirements.
+	// A successful authentication requires all of them being satisfied.
+	All []AuthenticationRequirement `yaml:"all,omitempty" json:"all,omitempty"`
+}
+
+// JWTAuthentication defines the JWT authentication
+type JWTAuthentication struct {
+	// Name of this JWT Provider, unique within the Proxy.
+	Name string `yaml:"name" json:"name"`
+
+	// JWT issuer ("iss" claim)
+	Issuer string `yaml:"issuer" json:"issuer"`
+
+	// A remote JWKS source
+	RemoteJWKS RemoteJWKS `yaml:"remote_jwks" json:"remote_jwks"`
+
+	// Audiences contains a list of audiences
+	Audiences []string `yaml:"audiences,omitempty" json:"audiences,omitempty"`
+
+	// Header name that will contain decoded JWT payload in requests forwarded to
+	// target.
+	ForwardPayloadHeader string `yaml:"forward_payload_header,omitempty" json:"forward_payload_header,omitempty"`
+
+	// Locations where JWT may be found. First match wins.
+	In []HTTPParameter `yaml:"in" json:"in"`
+}
+
+// RemoteJWKS contains information for remote JWKS
+type RemoteJWKS struct {
+	// URL of the JWKS
+	URL string `yaml:"url" json:"url"`
+
+	// CacheDuration of the JWKS
+	CacheDuration time.Duration `yaml:"cache_duration,omitempty" json:"cache_duration,omitempty"`
+}
+
+// ConsumerAuthorization is the configuration of API consumer authorization
+type ConsumerAuthorization struct {
+	FailOpen bool `yaml:"fail_open,omitempty" json:"fail_open,omitempty"`
+
+	// Locations of API consumer credential (API Key). First match wins.
+	In []HTTPParameter `yaml:"in" json:"in"`
+}
+
+// HTTPMatch is an HTTP request matching rule
+type HTTPMatch struct {
```

Comment:
What about `path_rewrite`? This can be done via Ext Authz IIUC via the `:path` pseudo-header, right?

Reply:
Can ext_authz rewrite headers?

Reply:
We can certainly send override headers, so it should be possible.

Reply:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto#envoy-v3-api-msg-service-auth-v3-okhttpresponse

Reply:
Ah, I see. So it is able to rewrite the header. What about reusing the StringModification here instead of defining two strings `path_rewrite` and `path_template`?

Reply:
Actually, I feel like if we're doing request modification that it should be a separate section...

Reply:
Yeah I have the same feeling. It should be another field in `Operation`?

Reply:
Yes. Maybe Target should be a section that contains request rewrite information?

Reply:
Based on other thread's naming suggestion, perhaps a "request_transformation" section?

Reply:
The name LGTM..but put `Target` under such a name seems a bit weird..

Reply:
No, I wouldn't put Target under that... I was thinking that Target would be a section that has a transformations section.

Reply:
I put a `RequestTransformation` under `Target`. It also feels like it could be put directly under `Operation`. I would be fine with either way.

Reply:
I agree it's not strictly a matching rule and is really a "transform." However, the problem with isolating the transform elsewhere is that `HTTPMatch` defines the _template_, so duplicating the template elsewhere can be error-prone and  confusing. Note that Envoy colocates the match rule and rewrite rule in the route config. Let's move this out of scope given the lack of clarity around the requirements.

Also let's definitely _not_ put it in `Target` when we do add it. Target is akin to a Cluster, while matching/transforming is correlated to Filter config (i.e. Router or HCM filters)